core/src/main/scala/kafka/log/Log.scala

```diff
@@ -428,15 +424,22 @@ class Log(@volatile var dir: File,
     }
   }
 
-  private def loadProducerState(lastOffset: Long): Unit = lock synchronized {
-    info(s"Loading producer state from offset $lastOffset for partition $topicPartition")
-
-    if (producerStateManager.latestSnapshotOffset.isEmpty) {
-      // if there are no snapshots to load producer state from, we assume that the brokers are
-      // being upgraded, which means there would be no previous idempotent/transactional producers
-      // to load state for. To avoid an expensive scan through all of the segments, we take
-      // empty snapshots from the start of the last two segments and the last offset. The purpose
-      // of taking the segment snapshots is to avoid the full scan in the case that the log needs
+  private def loadProducerState(lastOffset: Long, reloadFromCleanShutdown: Boolean): Unit = lock synchronized {
+    val messageFormatVersion = config.messageFormatVersion.messageFormatVersion
+    info(s"Loading producer state from offset $lastOffset for partition $topicPartition with message " +
+      s"format version $messageFormatVersion")
+
+    // We want to avoid unnecessary scanning of the log to build the producer state when the broker is being upgraded.
+    // There are two common cases where we can skip loading and write a new snapshot at the log end offset:
+    //
+    // 1. The broker has been upgraded, but the topic is still on the old message format.
+    // 2. The broker has been upgraded, the topic is on the new message format, and we had a clean shutdown.
```

Comment:
This is true because we should always have a snapshot file on a clean shutdown with the new message format, correct? In other words, are we assuming that if there isn't a snapshot file and the message format is new and we had a clean shutdown, that means this is an upgrade?